Feature solar elevation validation

seulgie opened the pull request:
Hi @peterdudfield ,

Based on your [feedback](https://github.com/openclimatefix/uk-pvnet-app/pull/221), I have 
- Changed the 'national_forecast_values' input in 'validate_forecast' to accept a 'pd.Series'
- Deleted 'forecast_data (dict)' and fixed 'lat' and 'lon' to (55.3781, -3.4360) 
- Added a 'try-except' block to handle cases where the 'time' column is missing in the dataset. In such cases, a default datetime index is generated.
- Updated the function call to 'validate_forecast' to reflect these changes.

Let me know if you need any further changes!

Best regards,
Seulgie

peterdudfield commented on src/pvnet_app/forecast_compiler.py:
```diff
@@ -97,6 +102,18 @@ def validate_forecast(
+    )
+
+    # Check if forecast values are > 0 when sun elevation > 10 degrees
+    elevation_above_10 = solpos["elevation"] > 10
```
could you make this `10` and environment variable that we can change. Potentially somelike FORECAST_VALIDATE_SUN_ELEVATION_LOWER_LIMIT`

peterdudfield commented on src/pvnet_app/forecast_compiler.py:
```diff
@@ -341,10 +358,23 @@ def compile_forecasts(self) -> None:
-        # Pass the entire national forecast array (for potential extra checks in future).
-        national_forecast_values = da_abs_national.sel(
-            output_label="forecast_mw", gsp_id=0).values
+        try:
```
could you remove this and just use `target_datetime_utc`

peterdudfield reviewed (changes requested):
Thanks so much for doing all this. Ive suggested a few changes, 

Could you add a test for this too?

seulgie commented:
Hi @peterdudfield, thanks for your detailed feedback !
I've made the requested updates to the forecast validation test:
- Sun elevation threshold is now retrieved from an environment variable
- Deleted the previous 'time' retrieval part and simply used self.valid_times
- The test has been integrated to `test_forecast_compiler.py` as part of the existing test suite.

Let me know if you need any further feedback 😃 
Thanks,
Seulgie

peterdudfield commented on tests/test_forecast_compiler.py:
```diff
@@ -125,3 +126,26 @@ def logger_func(message):
+
+    forecast_values =pd.Series([0 if elev < sun_elevation_limit else 50 for elev in sun_elevations], index=time_range)
+
+    # Mocking pvlib.solarposition.get_solarposition()
```
do you need to mock this?

peterdudfield replied:
might be good to test it out with the actual pvlib library

seulgie replied:
yes, I deleted mock and tested it out with the pvlib library.

peterdudfield commented on tests/test_forecast_compiler.py:
```diff
@@ -125,3 +125,35 @@ def logger_func(message):
+    are ≤ 0 while sun elevation is above SUN_ELEVATION_LOWER_LIMIT.
+    """
+    # Get Sun Elevation Limit value from the environment variable (default 10)
+    sun_elevation_limit = float(os.getenv("SUN_ELEVATION_LOWER_LIMIT", 10))
```
does this mean to be FORECAST_VALIDATE_SUN_ELEVATION_LOWER_LIMIT

peterdudfield replied:
ah could you add this env var in the readme.md. You might need to pull latest branch in? Thanks so much

peterdudfield reviewed (changes requested):
This looks great. I think its nearly there, just one small comment. 

Could you confirm the all the other tests run? Sorry they take some time

peterdudfield commented on src/pvnet_app/forecast_compiler.py:
```diff
@@ -90,19 +90,18 @@ def validate_forecast(
-                                          < -zig_zag_gap_error)  # Up then down by 500 MW
+    # Detect large fluctuations (either up or down by 250 MW)
+    large_jumps = (diff.abs() > zig_zag_gap_warning)
+    critical_jumps = (diff.abs() > zig_zag_gap_error)
```
im not sure this logic now works?, this justy looks for big jumps, not up and downs

seulgie replied:
Currently, the logic only detects large jumps (up or down) in the forecast values, but it doesn't specifically look for "up and down" patterns (i.e., a significant increase followed by a decrease). In fact, this behavior was identified while working on the elevation test, and I updated the code to account for this pattern.

Yes, it will be needed to adjust the implementation to specifically identify "up and down" fluctuations, which will improve the detection of these kinds of errors. Since this part was not directly related to this elevation logic ticket, I let it like this.

Let me know if you need further clarification or suggestions for this change 💡

peterdudfield replied:
But the old code that was remove did do this.

peterdudfield replied:
Yea, we specifically want to track up and then down fluctuations. To be honest we want to track, up, down, up fluctuations, but that is something extra

seulgie replied:
@peterdudfield , based on your feedback :
- All checks and tests except for the elevation part remain the same a
![test1](https://github.com/user-attachments/assets/1cac5745-c9ae-4140-b970-f4bbe3fb47e6)
s the current main code.
- The elevation test has successfully passed.

seulgie commented:
![test1](https://github.com/user-attachments/assets/0b83eb08-50fe-4d24-a333-e40eac8462a1)
I've made the following updates in the forecast validation logic as part of the elevation test ticket:
1. Datetime index mismatch: Fixed the issue where the `national_forecast_values` and solar position data had mismatched datetime indices.
2. Consistency in environment variable names: Standardized the environment variable name for the sun elevation threshold.
3. Test update: Updated the `test_validate_forecast_sun_elevation_check` to handle the corrected datetime indexing.
4. Fluctuation threshold correction: While testing on the `test_forecast_compiler`, I noticed some issues in the fluctuation threshold validation part, so I also corrected those in the test data to ensure the correct fluctuations are being tested and debug pytest errors.

Additionally, I ran all 8 tests locally in `test_forecast_compiler.py` and they all passed successfully, including the ones related to fluctuation validation that were previously failing. Please review the changes, and let me know yif you have any feedback.
Thanks!

peterdudfield commented:
This is linked to https://github.com/openclimatefix/uk-pvnet-app/issues/200

peterdudfield commented:
Thanks so much for this, ill probably merge it to a dev branch, to check all tests are wokring if thats ok. 
Thank you

seulgie commented:
That sounds great :) I really appreciate your detailed feedback—it helped me learn a lot throughout this ticket. 
Looking forward to working on the next one, thanks!

peterdudfield commented:
Hi @seulgie , would you be able to resolve the conflicts. 

Sorry we just moved validation to another file. 

I'm happy to resolve them if you prefer

seulgie commented:
Hi @peterdudfield , sure I will try to resolve the conflicts on Monday (2/25), before 17:00 (CET).

peterdudfield commented:
> Hi @peterdudfield , sure I will try to resolve the conflicts on Monday (2/25), before 17:00 (CET).

Cool, no problem, Thanks for letting me know. No pressure, do it when you can

seulgie commented:
Based on recent changes to move `validate_forecast` logic from `forecast_compiler.py` to new file `validate_forecast.py`,
ensured the integrity of the forecast validation flow by integrating the elevation checks from the new file.
Please review and let me know if you have any questions.

peterdudfield commented:
It still looks like there are merge conflicts? Cna you resolve? No worries if you cant, im happy to

peterdudfield commented:
Sure, ill try to merge onto a branch, and then do some merge conflicts
